probe-rs/src/probe/daplink/tools.rs

```diff
@@ -160,33 +172,56 @@ pub fn open_device_from_info(info: &DebugProbeInfo) -> Option<DAPLinkDevice> {
             // multiple open handles are not allowed on Windows.
             drop(handle);
 
-            if d_desc.vendor_id() == info.vendor_id
-                && d_desc.product_id() == info.product_id
-                && sn_str == info.serial_number
+            if d_desc.vendor_id() == selector.vendor_id
+                && d_desc.product_id() == selector.product_id
+                && sn_str == selector.serial_number
             {
                 // If the VID, PID, and potentially SN all match,
                 // attempt to open the device in v2 mode.
                 if let Some(device) = open_v2_device(device) {
-                    return Some(device);
+                    return Ok(device);
+                } else {
+                    // If we are here the device didn't support v2, try using hidapi to open in v1 mode.
+                    // If this doesn't work we give up and return None.
```

Comment:
I don't understand why you've duplicated the logic from below the `if let`, compared to the previous structure where if the loop finishes without hitting the `return`, the code below to open in HID mode will still run. The previous code already opened devices by serial number the same as this new block I believe.

Reply:
In the old code, if you have a jlink/stlink, and you would try to open a daplink, this would iterate through all usb devices and find no matching one, which ofc is correct because we aint trying to open a daplink! but after the loop it would fall back to try and open a dap hid device, which can open a wrong device :) I made both changes you commented because I experienced issues when testing the serial id feature. maybe i ficed the issue at the wrong location :) not saying i am in the right here ;) also, i am not sure whether our opening mechanism is the cleanest ;)